### PR TITLE
drivers: can: st: improve clock control

### PR DESCRIPTION
**Pull request by gmarull**
This PR improves the STM32 CAN driver by using the clock control instead of HAL. We should avoid using hard-coded values in HALs, as they bypass Devicetree as a source of hardware description. Also, this practice tends to make code less portable.

Note that there's still one hardcoded register access, `FDCAN_CONFIG`, it can be fixed later.

Found this _issue_ while working on <soc.h> cleanups.

### CODE REVIEW

File: drivers/can/can_stm32fd.c [hunk 1]
```diff
@@ -97,7 +104,12 @@ static int can_stm32fd_init(const struct device *dev)
 		return ret;
 	}
 
-	can_stm32fd_clock_enable();
+	ret = can_stm32fd_clock_enable(dev);
+	if (ret < 0) {
+		LOG_ERR("Could not turn CAN clock (%d)", ret);
```
Comment (martinjaeger) [inline drivers/can/can_stm32fd.c:109]: Nits
```suggestion
		LOG_ERR("Could not turn on CAN clock (%d)", ret);
```

**Review by martinjaeger — DISMISSED** [review]

File: drivers/can/can_stm32fd.c (hunk 1 above)
Comment (gmarull) [reply drivers/can/can_stm32fd.c]: fixed thanks

**Review by martinjaeger — DISMISSED** [review]

**Review by henrikbrixandersen — DISMISSED** [review]
Thanks!

**Review by alexanderwachter — DISMISSED** [review]

**Comment by carlescufi** [comment]
@erwango ping

**Review by erwango — CHANGES_REQUESTED** [review]
Please set clocks prop as `required: true` in `st,stm32-fdcan.yaml`.
Also, there's one more fdcan node that would require a clocks prop in `dts/arm/st/g4/stm32g473.dtsi`.
Otherwise thanks for doing this change!

**Comment by gmarull** [comment]
> Please set clocks prop as `required: true` in `st,stm32-fdcan.yaml`. Also, there's one more fdcan node that would require a clocks prop in `dts/arm/st/g4/stm32g473.dtsi`. Otherwise thanks for doing this change!

done

**Review by martinjaeger — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Pull request merged**